<a href="https://colab.research.google.com/github/Angelasilva15/inform-tica/blob/main/API1atualizado.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Task
Analyze the data in the "NCM.csv" file, filter the data in the "NO_NCM_POR" column that refers to the aviation and aeronautics sector, and generate an Excel spreadsheet named "NCMsetor_aeronautico".

## Carregar os dados

### Subtask:
Carregar o arquivo `NCM.csv` em um DataFrame do pandas.


## Salvar em excel

### Subtask:
Salvar o DataFrame filtrado em um novo arquivo Excel chamado `NCMsetor_aeronautico`.

**Reasoning**:
Save the filtered DataFrame to an Excel file without the index.

## Realizar cruzamento de dados (UF_MUN)

### Subtask:
Cruzar o DataFrame resultante com o DataFrame de UF_MUN para substituir os códigos de município pelos nomes.

**Reasoning**:
Merge the DataFrame with the UF_MUN DataFrame to replace municipality codes with names. Use a left merge to keep all rows from the export data DataFrame. The merge will be based on the municipality code columns.

In [ ]:
df_exp_aeronautico_sp = pd.merge(
    df_exp_aeronautico_sp,
    df_uf_mun[['CO_MUN_GEO', 'NO_MUN']],
    left_on='CO_MUN',
    right_on='CO_MUN_GEO',
    how='left'
)

# Drop the redundant 'CO_MUN_GEO' column after merging
df_exp_aeronautico_sp = df_exp_aeronautico_sp.drop('CO_MUN_GEO', axis=1)

display(df_exp_aeronautico_sp.head())
print(df_exp_aeronautico_sp.columns)

CO_ANO  CO_MES   SH4  CO_PAIS SG_UF_MUN   CO_MUN  KG_LIQUIDO    VL_FOB  \
0    2021       9  8802      249        SP  3416853        9973  14903100   
1    2021       8  8802       69        SP  3425904        2245    480000   
2    2021       7  8805      249        SP  3418800           7      5300   
3    2021       9  8802      767        SP  3443402           1      5966   
4    2021      11  8802      573        SP  3449904       33686  33229253   

                   NO_PAIS               NO_MUN  
0           Estados Unidos       GAVIAO PEIXOTO  
1                Austrália              JUNDIAI  
2           Estados Unidos            GUARULHOS  
3                    Suíça       RIBEIRAO PRETO  
4  Países Baixos (Holanda)  SAO JOSE DOS CAMPOS

Index(['CO_ANO', 'CO_MES', 'SH4', 'CO_PAIS', 'SG_UF_MUN', 'CO_MUN',
       'KG_LIQUIDO', 'VL_FOB', 'NO_PAIS', 'NO_MUN'],
      dtype='object')


## Salvar em Excel com abas por ano

### Subtask:
Salvar o DataFrame filtrado com os dados de exportação do setor aeronáutico (`df_exp_aeronautico`) em um novo arquivo Excel, com cada ano em uma aba separada.

**Reasoning**:
Create an Excel writer object and iterate through the unique years in the filtered DataFrame. For each year, filter the DataFrame to get the data for that year and save it to a separate sheet in the Excel file named after the year.

In [ ]:
excel_file_name = 'EXP_Aeronautico_por_Ano.xlsx'
with pd.ExcelWriter(excel_file_name) as writer:
    for year in df_exp_aeronautico['CO_ANO'].unique():
        df_year = df_exp_aeronautico[df_exp_aeronautico['CO_ANO'] == year]
        df_year.to_excel(writer, sheet_name=str(year), index=False)

print(f"Planilha '{excel_file_name}' gerada com sucesso, com dados separados por ano em abas.")

Planilha 'EXP_Aeronautico_por_Ano.xlsx' gerada com sucesso, com dados separados por ano em abas.


## Realizar cruzamento de dados (PAIS)

### Subtask:
Cruzar o DataFrame filtrado (`df_exp_aeronautico_sp`) com o DataFrame de países (`df_paises`) para substituir os códigos de país pelos nomes.

**Reasoning**:
Merge the filtered export data DataFrame with the countries DataFrame to replace country codes with names. Use a left merge to keep all rows from the filtered export data.

In [ ]:
df_exp_aeronautico_sp = pd.merge(
    df_exp_aeronautico_sp,
    df_paises[['CO_PAIS', 'NO_PAIS']],
    on='CO_PAIS',
    how='left'
)
display(df_exp_aeronautico_sp.head())
print(df_exp_aeronautico_sp.columns)

CO_ANO  CO_MES   SH4  CO_PAIS SG_UF_MUN   CO_MUN  KG_LIQUIDO    VL_FOB  \
0    2021       9  8802      249        SP  3416853        9973  14903100   
1    2021       8  8802       69        SP  3425904        2245    480000   
2    2021       7  8805      249        SP  3418800           7      5300   
3    2021       9  8802      767        SP  3443402           1      5966   
4    2021      11  8802      573        SP  3449904       33686  33229253   

                   NO_PAIS  
0           Estados Unidos  
1                Austrália  
2           Estados Unidos  
3                    Suíça  
4  Países Baixos (Holanda)

Index(['CO_ANO', 'CO_MES', 'SH4', 'CO_PAIS', 'SG_UF_MUN', 'CO_MUN',
       'KG_LIQUIDO', 'VL_FOB', 'NO_PAIS'],
      dtype='object')


## Carregar dados de referência para cruzamento

### Subtask:
Carregar os arquivos `/content/PAIS.csv`, `/content/UF_MUN.csv` e `/content/UF_MUN_SP.xlsx` em DataFrames do pandas.

**Reasoning**:
Load the specified CSV and Excel files into pandas DataFrames and display their heads and columns to verify the loading.

In [ ]:
df_paises = pd.read_csv('/content/PAIS.csv', encoding='latin1', sep=';')
display(df_paises.head())
print(df_paises.columns)

df_uf_mun = pd.read_csv('/content/UF_MUN.csv', encoding='latin1', sep=';')
display(df_uf_mun.head())
print(df_uf_mun.columns)

df_uf_mun_sp_ref = pd.read_excel('/content/UF_MUN_SP.xlsx')
display(df_uf_mun_sp_ref.head())
print(df_uf_mun_sp_ref.columns)

CO_PAIS  CO_PAIS_ISON3 CO_PAIS_ISOA3                 NO_PAIS  \
0        0            898           ZZZ            Não Definido   
1       13              4           AFG             Afeganistão   
2       15            248           ALA            Aland, Ilhas   
3       17              8           ALB                 Albânia   
4       20            724           ESP  Alboran-Perejil, Ilhas   

                NO_PAIS_ING             NO_PAIS_ESP  
0               Not defined             No definido  
1               Afghanistan              Afganistan  
2             Aland Islands            Alans, Islas  
3                   Albania                 Albania  
4  Alboran-Perejil, Islands  Alboran-Perejil, Islas

Index(['CO_PAIS', 'CO_PAIS_ISON3', 'CO_PAIS_ISOA3', 'NO_PAIS', 'NO_PAIS_ING',
       'NO_PAIS_ESP'],
      dtype='object')


CO_MUN_GEO                  NO_MUN              NO_MUN_MIN SG_UF
0     2502003  BELEM DO BREJO DO CRUZ  Belém do Brejo do Cruz    PB
1     2502052      BERNARDINO BATISTA      Bernardino Batista    PB
2     2502102             BOA VENTURA             Boa Ventura    PB
3     2502151               BOA VISTA               Boa Vista    PB
4     2502201               BOM JESUS               Bom Jesus    PB

Index(['CO_MUN_GEO', 'NO_MUN', 'NO_MUN_MIN', 'SG_UF'], dtype='object')


CO_MUN_GEO                  NO_MUN              NO_MUN_MIN SG_UF
0     3446900             SANTA LUCIA             Santa Lúcia    SP
1     3447007    SANTA MARIA DA SERRA    Santa Maria da Serra    SP
2     3447106          SANTA MERCEDES          Santa Mercedes    SP
3     3447205  SANTANA DA PONTE PENSA  Santana da Ponte Pensa    SP
4     3447304     SANTANA DE PARNAIBA     Santana de Parnaíba    SP

Index(['CO_MUN_GEO', 'NO_MUN', 'NO_MUN_MIN', 'SG_UF'], dtype='object')


## Filtrar por estado

### Subtask:
Filtrar o DataFrame consolidado para incluir apenas as linhas onde a coluna `SG_UF_MUN` é igual a 'SP'.

**Reasoning**:
Filter the DataFrame to include only rows where 'SG_UF_MUN' is 'SP' and display the head of the filtered DataFrame.

In [ ]:
df_exp_aeronautico_sp = df_exp_aeronautico_consolidado[df_exp_aeronautico_consolidado['SG_UF_MUN'] == 'SP']
display(df_exp_aeronautico_sp.head())

CO_ANO  CO_MES   SH4  CO_PAIS SG_UF_MUN   CO_MUN  KG_LIQUIDO    VL_FOB
0    2021       9  8802      249        SP  3416853        9973  14903100
1    2021       8  8802       69        SP  3425904        2245    480000
3    2021       7  8805      249        SP  3418800           7      5300
4    2021       9  8802      767        SP  3443402           1      5966
5    2021      11  8802      573        SP  3449904       33686  33229253

## Carregar os dados consolidados

### Subtask:
Carregar o arquivo `/content/EXP_Aeronautico_NCM_consolidado.xlsx` em um DataFrame do pandas.

**Reasoning**:
Import the pandas library and load the Excel file into a DataFrame. Then display the head of the DataFrame and its columns to verify the data loading.

In [ ]:
import pandas as pd

df_exp_aeronautico_consolidado = pd.read_excel('/content/EXP_Aeronautico_NCM_consolidado.xlsx')
display(df_exp_aeronautico_consolidado.head())
print(df_exp_aeronautico_consolidado.columns)

CO_ANO  CO_MES   SH4  CO_PAIS SG_UF_MUN   CO_MUN  KG_LIQUIDO    VL_FOB
0    2021       9  8802      249        SP  3416853        9973  14903100
1    2021       8  8802       69        SP  3425904        2245    480000
2    2021       1  8802      589        AC  1200252          90      2444
3    2021       7  8805      249        SP  3418800           7      5300
4    2021       9  8802      767        SP  3443402           1      5966

Index(['CO_ANO', 'CO_MES', 'SH4', 'CO_PAIS', 'SG_UF_MUN', 'CO_MUN',
       'KG_LIQUIDO', 'VL_FOB'],
      dtype='object')


## Salvar em excel

### Subtask:
Salvar o DataFrame filtrado em um novo arquivo Excel chamado `NCMsetor_aeronautico.xlsx`.

**Reasoning**:
Save the filtered DataFrame to an Excel file without the index.

In [ ]:
df_ncm_aeronautico.to_excel('NCMsetor_aeronautico.xlsx', index=False)

## Filtrar os dados

### Subtask:
Filtrar o DataFrame para incluir apenas as linhas onde a coluna `NO_NCM_POR` contém termos relacionados aos setores de aviação e aeronáutica.

**Reasoning**:
Filter the DataFrame `df_ncm` based on keywords related to aviation and aeronautics in the `NO_NCM_POR` column, performing a case-insensitive search.

In [ ]:
aviation_keywords = ['aviação', 'aeronáutica', 'aeronave', 'avião', 'helicóptero', 'motor de aviação', 'peças de aeronaves']
filter_pattern = '|'.join(aviation_keywords)
df_ncm_aeronautico = df_ncm[df_ncm['NO_NCM_POR'].str.contains(filter_pattern, case=False, na=False)]
display(df_ncm_aeronautico.head())

CO_NCM  CO_UNID  CO_SH6  CO_PPE  CO_PPI  CO_FAT_AGREG CO_CUCI_ITEM  \
538   88069100       11  880691     NaN     NaN           NaN         7922   
542   88061000       11  880610     NaN     NaN           NaN         7922   
1245  99980102       10  999801  5020.0  4999.0           5.0         9310   
1540  84071000       11  840710  3191.0  3191.0           3.0        71311   
2637  84798931       11  847989  3990.0  3990.0           3.0        72849   

      CO_CGCE_N3  CO_SIIT  CO_ISIC_CLASSE  CO_EXP_SUBSET  \
538          110      NaN            2829            NaN   
542          110      NaN            2829            NaN   
1245         421   9000.0            8999          602.0   
1540         280   1000.0            3030         1014.0   
2637         280   2000.0            2829          999.0   

                                             NO_NCM_POR  \
538   Outros veículos aéreos (aeronaves) não tripula...   
542   Veículos aéreos (aeronaves) não tripulados, co...   
1245  Consumo de bordo - combustíveis e lubrificante...   
1540                  Motores de explosão, para aviação   
2637  Limpadores de pára-brisas elétricos, para aero...   

                                             NO_NCM_ESP  \
538   Otros vehículos aéreos no tripulados (aeronave...   
542   Vehículos aéreos no tripulados de transporte d...   
1245  Consumo de bordo - combustibles y lubric.p/aer...   
1540                  Motores de explosión, de aviación   
2637           Limpiaparabrisas eléctricos, p/aeronaves   

                                             NO_NCM_ING  
538   Other unmanned aerial vehicles (aircraft), of ...  
542   Unmanned aerial vehicles (aircraft), designed ...  
1245  Onboard consump.fuels and lubrication f/aircrafts  
1540                   Combustion engines, for aviation  
2637          Electric windshield wipers, for aircrafts

**Reasoning**:
The previous attempt to read the CSV with specified encoding, delimiter, and quoting parameters still resulted in a ParserError, indicating an inconsistent number of fields in different rows. This often happens with messy CSV files. I will try reading the file again, keeping the 'ISO-8859-1' encoding, semicolon delimiter, and 'python' engine, but this time, I will use `on_bad_lines='skip'` to skip the problematic lines during the parsing process.

In [ ]:
df_ncm = pd.read_csv('NCM.csv', encoding='ISO-8859-1', sep=';', engine='python', on_bad_lines='skip')
display(df_ncm.head())

CO_NCM  CO_UNID  CO_SH6  CO_PPE  CO_PPI  CO_FAT_AGREG CO_CUCI_ITEM  \
0  85284110       11  852841  3145.0  3145.0           3.0         7613   
1  85423210       11  854232  3235.0  3235.0           3.0        77644   
2  85423221       11  854232  3235.0  3235.0           3.0        77644   
3  85423229       11  854232  3235.0  3235.0           3.0        77644   
4  85423291       11  854232  3235.0  3235.0           3.0        77644   

   CO_CGCE_N3  CO_SIIT  CO_ISIC_CLASSE  CO_EXP_SUBSET  \
0         110   1000.0            2620          902.0   
1         270   1000.0            2610         1106.0   
2         270   1000.0            2610         1106.0   
3         270   1000.0            2610         1106.0   
4         270   1000.0            2610         1106.0   

                                          NO_NCM_POR  \
0  Monitores com tubo de raios catódicos, dos tip...   
1                              Memórias não montadas   
2  Memórias, montadas, próprias para montagem em ...   
3                  Outras memórias digitais montadas   
4  Outras memórias dos tipos RAM estáticas (SRAM)...   

                                          NO_NCM_ESP  \
0  Unid. de salida p/video, c/ tubo de rayos cató...   
1          Circuitos integr.monolít.digit.sin montar   
2  Memórias ram <=25ns, eprom, eeprom, prom, rom,...   
3                  Otras memorias digitales montadas   
4  Otras memórias ram <25ns, eprom, eeprom, prom,...   

                                          NO_NCM_ING  
0  Output unit for video, with monocr. crt (catho...  
1   Digit.monolithic integrated circuits, no mounted  
2    Ram memories <=25ns,eprom,eeprom,prom,rom,flash  
3                     Other digital digital memories  
4   Other ram memories <25ns,eprom,eeprom,prom,flash

**Reasoning**:
The previous attempt with 'ISO-8859-1' encoding and semicolon delimiter with the 'python' engine still resulted in a ParserError related to quotes and semicolons. This suggests there might be issues with how fields are quoted in the CSV. I will try reading the file again, keeping the 'ISO-8859-1' encoding and semicolon delimiter, but also explicitly setting the `quotechar` to '"' and the `quoting` parameter to `csv.QUOTE_NONE` to handle potential issues with quoted fields. I will also keep the 'python' engine.

In [ ]:
import csv
df_ncm = pd.read_csv('NCM.csv', encoding='ISO-8859-1', sep=';', engine='python', quotechar='"', quoting=csv.QUOTE_NONE)
display(df_ncm.head())

ParserError: Expected 14 fields in line 210, saw 15

**Reasoning**:
The previous attempt to read the CSV with 'latin1' encoding resulted in a ParserError, indicating an issue with the delimiter or structure of the file. I will try reading the file again, but this time I will explicitly set the delimiter to a semicolon, which is a common delimiter in CSV-like files, and also try a different encoding like 'ISO-8859-1' which is similar to 'latin1'. I will also specify the 'engine' as 'python' to handle potential parsing issues more robustly.

In [ ]:
df_ncm = pd.read_csv('NCM.csv', encoding='ISO-8859-1', sep=';', engine='python')
display(df_ncm.head())

ParserError: ';' expected after '"'

**Reasoning**:
The previous attempt to read the CSV failed due to a UnicodeDecodeError. This suggests that the file is not encoded in UTF-8. I will try reading the file again, specifying a different encoding that is common for such errors, such as 'latin1'.

In [ ]:
df_ncm = pd.read_csv('NCM.csv', encoding='latin1')
display(df_ncm.head())

ParserError: Error tokenizing data. C error: Expected 5 fields in line 4, saw 18


**Reasoning**:
Import the pandas library and load the NCM.csv file into a DataFrame. Then display the head of the DataFrame to verify the data loading.



In [ ]:
import pandas as pd

df_ncm = pd.read_csv('NCM.csv')
display(df_ncm.head())

UnicodeDecodeError: 'utf-8' codec can't decode byte 0xf3 in position 282: invalid continuation byte

**Reasoning**:
The previous attempt to read the CSV failed due to a UnicodeDecodeError. This suggests that the file is not encoded in UTF-8. I will try reading the file again, specifying a different encoding that is common for such errors, such as 'latin1'.



In [ ]:
df_ncm = pd.read_csv('NCM.csv', encoding='latin1')
display(df_ncm.head())

ParserError: Error tokenizing data. C error: Expected 5 fields in line 4, saw 18


**Reasoning**:
The previous attempt to read the CSV with 'latin1' encoding resulted in a ParserError, indicating an issue with the delimiter or structure of the file. I will try reading the file again, but this time I will explicitly set the delimiter to a semicolon, which is a common delimiter in CSV-like files, and also try a different encoding like 'ISO-8859-1' which is similar to 'latin1'. I will also specify the 'engine' as 'python' to handle potential parsing issues more robustly.



In [ ]:
df_ncm = pd.read_csv('NCM.csv', encoding='ISO-8859-1', sep=';', engine='python')
display(df_ncm.head())

ParserError: ';' expected after '"'

**Reasoning**:
The previous attempt with 'ISO-8859-1' encoding and semicolon delimiter with the 'python' engine still resulted in a ParserError related to quotes and semicolons. This suggests there might be issues with how fields are quoted in the CSV. I will try reading the file again, keeping the 'ISO-8859-1' encoding and semicolon delimiter, but also explicitly setting the `quotechar` to '"' and the `quoting` parameter to `csv.QUOTE_NONE` to handle potential issues with quoted fields. I will also keep the 'python' engine.



In [ ]:
import csv
df_ncm = pd.read_csv('NCM.csv', encoding='ISO-8859-1', sep=';', engine='python', quotechar='"', quoting=csv.QUOTE_NONE)
display(df_ncm.head())

ParserError: Expected 14 fields in line 210, saw 15

**Reasoning**:
The previous attempt to read the CSV with specified encoding, delimiter, and quoting parameters still resulted in a ParserError, indicating an inconsistent number of fields in different rows. This often happens with messy CSV files. I will try reading the file again, keeping the 'ISO-8859-1' encoding, semicolon delimiter, and 'python' engine, but this time, I will use `on_bad_lines='skip'` to skip the problematic lines during the parsing process.



In [ ]:
df_ncm = pd.read_csv('NCM.csv', encoding='ISO-8859-1', sep=';', engine='python', on_bad_lines='skip')
display(df_ncm.head())

CO_NCM  CO_UNID  CO_SH6  CO_PPE  CO_PPI  CO_FAT_AGREG CO_CUCI_ITEM  \
0  85284110       11  852841  3145.0  3145.0           3.0         7613   
1  85423210       11  854232  3235.0  3235.0           3.0        77644   
2  85423221       11  854232  3235.0  3235.0           3.0        77644   
3  85423229       11  854232  3235.0  3235.0           3.0        77644   
4  85423291       11  854232  3235.0  3235.0           3.0        77644   

   CO_CGCE_N3  CO_SIIT  CO_ISIC_CLASSE  CO_EXP_SUBSET  \
0         110   1000.0            2620          902.0   
1         270   1000.0            2610         1106.0   
2         270   1000.0            2610         1106.0   
3         270   1000.0            2610         1106.0   
4         270   1000.0            2610         1106.0   

                                          NO_NCM_POR  \
0  Monitores com tubo de raios catódicos, dos tip...   
1                              Memórias não montadas   
2  Memórias, montadas, próprias para montagem em ...   
3                  Outras memórias digitais montadas   
4  Outras memórias dos tipos RAM estáticas (SRAM)...   

                                          NO_NCM_ESP  \
0  Unid. de salida p/video, c/ tubo de rayos cató...   
1          Circuitos integr.monolít.digit.sin montar   
2  Memórias ram <=25ns, eprom, eeprom, prom, rom,...   
3                  Otras memorias digitales montadas   
4  Otras memórias ram <25ns, eprom, eeprom, prom,...   

                                          NO_NCM_ING  
0  Output unit for video, with monocr. crt (catho...  
1   Digit.monolithic integrated circuits, no mounted  
2    Ram memories <=25ns,eprom,eeprom,prom,rom,flash  
3                     Other digital digital memories  
4   Other ram memories <25ns,eprom,eeprom,prom,flash

## Filtrar os dados

### Subtask:
Filtrar o DataFrame para incluir apenas as linhas onde a coluna `NO_NCM_POR` contém termos relacionados aos setores de aviação e aeronáutica.


**Reasoning**:
Filter the DataFrame `df_ncm` based on keywords related to aviation and aeronautics in the `NO_NCM_POR` column, performing a case-insensitive search.



In [ ]:
aviation_keywords = ['aviação', 'aeronáutica', 'aeronave', 'avião', 'helicóptero', 'motor de aviação', 'peças de aeronaves']
filter_pattern = '|'.join(aviation_keywords)
df_ncm_aeronautico = df_ncm[df_ncm['NO_NCM_POR'].str.contains(filter_pattern, case=False, na=False)]
display(df_ncm_aeronautico.head())

CO_NCM  CO_UNID  CO_SH6  CO_PPE  CO_PPI  CO_FAT_AGREG CO_CUCI_ITEM  \
538   88069100       11  880691     NaN     NaN           NaN         7922   
542   88061000       11  880610     NaN     NaN           NaN         7922   
1245  99980102       10  999801  5020.0  4999.0           5.0         9310   
1540  84071000       11  840710  3191.0  3191.0           3.0        71311   
2637  84798931       11  847989  3990.0  3990.0           3.0        72849   

      CO_CGCE_N3  CO_SIIT  CO_ISIC_CLASSE  CO_EXP_SUBSET  \
538          110      NaN            2829            NaN   
542          110      NaN            2829            NaN   
1245         421   9000.0            8999          602.0   
1540         280   1000.0            3030         1014.0   
2637         280   2000.0            2829          999.0   

                                             NO_NCM_POR  \
538   Outros veículos aéreos (aeronaves) não tripula...   
542   Veículos aéreos (aeronaves) não tripulados, co...   
1245  Consumo de bordo - combustíveis e lubrificante...   
1540                  Motores de explosão, para aviação   
2637  Limpadores de pára-brisas elétricos, para aero...   

                                             NO_NCM_ESP  \
538   Otros vehículos aéreos no tripulados (aeronave...   
542   Vehículos aéreos no tripulados de transporte d...   
1245  Consumo de bordo - combustibles y lubric.p/aer...   
1540                  Motores de explosión, de aviación   
2637           Limpiaparabrisas eléctricos, p/aeronaves   

                                             NO_NCM_ING  
538   Other unmanned aerial vehicles (aircraft), of ...  
542   Unmanned aerial vehicles (aircraft), designed ...  
1245  Onboard consump.fuels and lubrication f/aircrafts  
1540                   Combustion engines, for aviation  
2637          Electric windshield wipers, for aircrafts

## Salvar em excel

### Subtask:
Salvar o DataFrame filtrado em um novo arquivo Excel chamado `NCMsetor_aeronautico.xlsx`.


**Reasoning**:
Save the filtered DataFrame to an Excel file without the index.



In [ ]:
df_ncm_aeronautico.to_excel('NCMsetor_aeronautico.xlsx', index=False)

## Summary:

### Data Analysis Key Findings

* The initial dataset from "NCM.csv" required specific parameters for successful loading due to encoding, delimiter, and parsing issues, ultimately being read with `encoding='ISO-8859-1'`, `sep=';'`, `engine='python'`, and `on_bad_lines='skip'`.
* The filtering process successfully identified and isolated data points related to the aviation and aeronautics sector by searching for a predefined list of Portuguese keywords within the `NO_NCM_POR` column, resulting in a filtered DataFrame named `df_ncm_aeronautico`.
* The final filtered dataset was successfully saved to an Excel file named "NCMsetor_aeronautico.xlsx", with the DataFrame index excluded from the output.

### Insights or Next Steps

* The filtered Excel file "NCMsetor_aeronautico.xlsx" now contains NCM codes and descriptions specifically relevant to the aviation and aeronautics industry, ready for further, more detailed analysis.
* A potential next step could involve analyzing the frequency or distribution of different NCM codes within the filtered aviation dataset to identify the most common types of goods or services categorized under aviation/aeronautics.


# Task
Analyze the data in the "UF_MUN.csv" file, filter the data in the "SG_UF" column to include only data referring to "SP", and display the filtered data.

## Carregar os dados

### Subtask:
Carregar o arquivo `UF_MUN.csv` em um DataFrame do pandas.


**Reasoning**:
Save the filtered DataFrame `df_uf_mun_sp` to an Excel file named 'UF_MUN_SP.xlsx' without including the index.

In [ ]:
df_uf_mun_sp.to_excel('UF_MUN_SP.xlsx', index=False)

## Salvar em Excel

### Subtask:
Salvar o DataFrame filtrado para o estado de São Paulo (`df_uf_mun_sp`) em um novo arquivo Excel.

## Salvar em excel

### Subtask:
Salvar o DataFrame filtrado (setor de aviação e aeronáutica) em um novo arquivo Excel chamado `UFsetor_aviação.xlsx`.

## Salvar dados de exportação filtrados em Excel

### Subtask:
Salvar o DataFrame filtrado com os dados de exportação do setor aeronáutico (`df_exp_aeronautico`) em um novo arquivo Excel.

## Salvar em excel

### Subtask:
Salvar o DataFrame filtrado (setor de aviação e aeronáutica) em um novo arquivo Excel chamado `UFsetor_aviação.xlsx`.

## Salvar dados de exportação filtrados em Excel

### Subtask:
Salvar o DataFrame filtrado com os dados de exportação do setor aeronáutico (`df_exp_aeronautico`) em um novo arquivo Excel.

## Salvar em excel

### Subtask:
Salvar o DataFrame filtrado (setor de aviação e aeronáutica) em um novo arquivo Excel chamado `UFsetor_aviação.xlsx`.

## Salvar dados de exportação filtrados em Excel

### Subtask:
Salvar o DataFrame filtrado com os dados de exportação do setor aeronáutico (`df_exp_aeronautico`) em um novo arquivo Excel.

**Reasoning**:
Save the filtered DataFrame `df_exp_aeronautico` to an Excel file named 'EXP_Aeronautico_2024.xlsx' without including the index.

In [ ]:
df_exp_aeronautico.to_excel('EXP_Aeronautico_2024.xlsx', index=False)

**Reasoning**:
Save the filtered DataFrame `df_ncm_aeronautico` to an Excel file named 'UFsetor_aviação.xlsx' without including the index.

In [ ]:
df_ncm_aeronautico.to_excel('UFsetor_aviação.xlsx', index=False)

**Reasoning**:
Import the pandas library and load the UF_MUN.csv file into a DataFrame. Then display the head of the DataFrame to verify the data loading.



In [ ]:
import pandas as pd

df_uf_mun = pd.read_csv('UF_MUN.csv')
display(df_uf_mun.head())

UnicodeDecodeError: 'utf-8' codec can't decode byte 0xe9 in position 83: invalid continuation byte

**Reasoning**:
The previous attempt to read the CSV failed due to a UnicodeDecodeError. This suggests that the file is not encoded in UTF-8. I will try reading the file again, specifying a different encoding that is common for such errors, such as 'latin1'.



In [ ]:
df_uf_mun = pd.read_csv('UF_MUN.csv', encoding='latin1')
display(df_uf_mun.head())

CO_MUN_GEO;"NO_MUN";"NO_MUN_MIN";"SG_UF"
0  2502003;"BELEM DO BREJO DO CRUZ";"Belém do Bre...
1  2502052;"BERNARDINO BATISTA";"Bernardino Batis...
2           2502102;"BOA VENTURA";"Boa Ventura";"PB"
3               2502151;"BOA VISTA";"Boa Vista";"PB"
4               2502201;"BOM JESUS";"Bom Jesus";"PB"

**Reasoning**:
The data was loaded, but it appears to be delimited by semicolons and potentially quoted. I need to read the CSV again, specifying the semicolon delimiter to correctly separate the columns.



In [ ]:
df_uf_mun = pd.read_csv('UF_MUN.csv', encoding='latin1', sep=';')
display(df_uf_mun.head())

CO_MUN_GEO                  NO_MUN              NO_MUN_MIN SG_UF
0     2502003  BELEM DO BREJO DO CRUZ  Belém do Brejo do Cruz    PB
1     2502052      BERNARDINO BATISTA      Bernardino Batista    PB
2     2502102             BOA VENTURA             Boa Ventura    PB
3     2502151               BOA VISTA               Boa Vista    PB
4     2502201               BOM JESUS               Bom Jesus    PB

## Filtrar os dados

### Subtask:
Filtrar o DataFrame para incluir apenas as linhas onde a coluna `SG_UF` é igual a 'SP'.


**Reasoning**:
Filter the `df_uf_mun` DataFrame to include only the rows where the 'SG_UF' column is equal to 'SP'.



In [ ]:
df_uf_mun_sp = df_uf_mun[df_uf_mun['SG_UF'] == 'SP']
display(df_uf_mun_sp.head())

CO_MUN_GEO                  NO_MUN              NO_MUN_MIN SG_UF
1170     3446900             SANTA LUCIA             Santa Lúcia    SP
1171     3447007    SANTA MARIA DA SERRA    Santa Maria da Serra    SP
1172     3447106          SANTA MERCEDES          Santa Mercedes    SP
1173     3447205  SANTANA DA PONTE PENSA  Santana da Ponte Pensa    SP
1174     3447304     SANTANA DE PARNAIBA     Santana de Parnaíba    SP

## Exibir os dados filtrados

### Subtask:
Mostrar as primeiras linhas do DataFrame filtrado para verificação.


**Reasoning**:
Display the first 5 rows of the filtered DataFrame to verify the result of the filtering operation.



In [ ]:
display(df_uf_mun_sp.head())

CO_MUN_GEO                  NO_MUN              NO_MUN_MIN SG_UF
1170     3446900             SANTA LUCIA             Santa Lúcia    SP
1171     3447007    SANTA MARIA DA SERRA    Santa Maria da Serra    SP
1172     3447106          SANTA MERCEDES          Santa Mercedes    SP
1173     3447205  SANTANA DA PONTE PENSA  Santana da Ponte Pensa    SP
1174     3447304     SANTANA DE PARNAIBA     Santana de Parnaíba    SP

## Summary:

### Data Analysis Key Findings

*   The data was successfully loaded from the "UF\_MUN.csv" file by specifying the 'latin1' encoding and using a semicolon (`;`) as the delimiter.
*   The data was successfully filtered to include only entries where the "SG\_UF" column is equal to 'SP'.
*   The first five rows of the filtered data were displayed, confirming that the filtering process was successful and only data for 'SP' was included.

### Insights or Next Steps

*   The filtered DataFrame `df_uf_mun_sp` is ready for further analysis or visualization specific to the state of São Paulo.
*   Further analysis could involve exploring other columns in the filtered data or joining this data with other relevant datasets.


# Task
Analise todos os dados do arquivo "/content/EXP_2024_MUN.csv", e faça um filtro separando somente os dados que forem iguais ao CO_SH4 do arquivo "/content/setor_aeronautico.xlsx".

## Carregar os dados de exportação

### Subtask:
Carregar o arquivo `/content/EXP_2024_MUN.csv` em um DataFrame do pandas.


**Reasoning**:
Import the pandas library and load the EXP_2024_MUN.csv file into a DataFrame. Then display the head of the DataFrame to verify the data loading.



In [ ]:
df_exp_2024_mun = pd.read_csv('/content/EXP_2024_MUN.csv')
display(df_exp_2024_mun.head())

CO_ANO;"CO_MES";"SH4";"CO_PAIS";"SG_UF_MUN";"CO_MUN";"KG_LIQUIDO";"VL_FOB"
0   2024;"02";"8708";"158";"MG";"3106705";3346;97359                        
1  2024;"09";"1516";"845";"SP";"3425904";349090;5...                        
2   2024;"07";"8536";"158";"SC";"4208906";1711;20706                        
3  2024;"08";"4010";"063";"SP";"3452205";4586;176026                        
4      2024;"12";"0701";"476";"SC";"4216206";330;274

**Reasoning**:
The previous attempt to read the CSV loaded the data into a single column, indicating a delimiter issue. I will read the CSV again, explicitly specifying the semicolon as the delimiter.



In [ ]:
df_exp_2024_mun = pd.read_csv('/content/EXP_2024_MUN.csv', sep=';')
display(df_exp_2024_mun.head())

CO_ANO  CO_MES   SH4  CO_PAIS SG_UF_MUN   CO_MUN  KG_LIQUIDO  VL_FOB
0    2024       2  8708      158        MG  3106705        3346   97359
1    2024       9  1516      845        SP  3425904      349090  541361
2    2024       7  8536      158        SC  4208906        1711   20706
3    2024       8  4010       63        SP  3452205        4586  176026
4    2024      12   701      476        SC  4216206         330     274

## Carregar os dados do setor aeronáutico

### Subtask:
Carregar o arquivo `/content/setor_aeronautico.xlsx` em um DataFrame do pandas e extrair a coluna `CO_SH4`.


**Reasoning**:
Import the pandas library and load the Excel file into a DataFrame, then extract the 'CO_SH4' column and display the head of the extracted data.



In [ ]:
import pandas as pd

df_setor_aeronautico = pd.read_excel('/content/setor_aeronautico.xlsx')
sh4_codes = df_setor_aeronautico['CO_SH4']
display(sh4_codes.head())

0    8801
1    8801
2    8801
3    8802
4    8802
Name: CO_SH4, dtype: int64

## Filtrar os dados de exportação

### Subtask:
Filtrar o DataFrame de exportação para incluir apenas as linhas onde o `CO_SH4` corresponde aos valores encontrados no arquivo do setor aeronáutico.


**Reasoning**:
Filter the export data DataFrame to include only the rows where the 'CO_SH4' column values are present in the 'sh4_codes' Series.



In [ ]:
df_exp_aeronautico = df_exp_2024_mun[df_exp_2024_mun['CO_SH4'].isin(sh4_codes)]
display(df_exp_aeronautico.head())

KeyError: 'CO_SH4'

**Reasoning**:
The previous attempt failed because the column 'CO_SH4' was not found in the DataFrame. I need to check the column names of `df_exp_2024_mun` to identify the correct column to use for filtering.



In [ ]:
print(df_exp_2024_mun.columns)

Index(['CO_ANO', 'CO_MES', 'SH4', 'CO_PAIS', 'SG_UF_MUN', 'CO_MUN',
       'KG_LIQUIDO', 'VL_FOB'],
      dtype='object')


**Reasoning**:
The column names were printed, and it appears the column representing the SH4 code is named 'SH4', not 'CO_SH4'. I will filter the DataFrame using the correct column name.



In [ ]:
df_exp_aeronautico = df_exp_2024_mun[df_exp_2024_mun['SH4'].isin(sh4_codes)]
display(df_exp_aeronautico.head())

CO_ANO  CO_MES   SH4  CO_PAIS SG_UF_MUN   CO_MUN  KG_LIQUIDO    VL_FOB
529       2024       5  8802      764        SP  3449904          93    199328
13816     2024       9  8802       63        SP  3450308         850     25000
48585     2024      10  8802      245        SP  3449904       67372  78178250
78677     2024      12  8802      249        SP  3421101         840    472000
109552    2024       8  8802      249        MG  3117876       30977  20000000

## Exibir os dados filtrados

### Subtask:
Mostrar as primeiras linhas do DataFrame filtrado para verificação.


**Reasoning**:
Display the first 5 rows of the filtered DataFrame to verify the result of the filtering operation.



In [ ]:
display(df_exp_aeronautico.head())

CO_ANO  CO_MES   SH4  CO_PAIS SG_UF_MUN   CO_MUN  KG_LIQUIDO    VL_FOB
529       2024       5  8802      764        SP  3449904          93    199328
13816     2024       9  8802       63        SP  3450308         850     25000
48585     2024      10  8802      245        SP  3449904       67372  78178250
78677     2024      12  8802      249        SP  3421101         840    472000
109552    2024       8  8802      249        MG  3117876       30977  20000000

## Summary:

### Data Analysis Key Findings

*   The export data from `/content/EXP_2024_MUN.csv` was successfully loaded and parsed using the semicolon (`;`) delimiter.
*   The aeronautical sector data from `/content/setor_aeronautico.xlsx` was loaded, and the `CO_SH4` column was extracted.
*   The export data DataFrame was filtered to include only entries where the `SH4` column matched the codes from the aeronautical sector file.
*   The filtered DataFrame `df_exp_aeronautico` contains export data related to the aeronautical sector, as confirmed by the presence of relevant `SH4` codes (e.g., 8802) in the displayed head.

### Insights or Next Steps

*   The filtered DataFrame can now be used for further analysis specific to the aeronautical sector's export performance in 2024.
*   Consider exploring the total export value (VL\_FOB) and quantity (KG\_LIQUIDO) for the aeronautical sector within this dataset.


# Task
Analise todos os dados dos arquivos: "/content/EXP_2021_MUN.csv", "/content/EXP_2022_MUN.csv", "/content/EXP_2023_MUN.csv", "/content/EXP_2024_MUN.csv", "/content/EXP_2025_MUN.csv", e faça um filtro separando somente os dados que forem iguais ao CO_SH4 do arquivo "/content/setor_aeronautico.xlsx".

## Carregar os dados do setor aeronáutico

### Subtask:
Carregar o arquivo `/content/setor_aeronautico.xlsx` em um DataFrame do pandas e extrair a coluna `CO_SH4`.


**Reasoning**:
Load the Excel file into a DataFrame and extract the 'CO_SH4' column.



In [ ]:
df_setor_aeronautico = pd.read_excel('/content/setor_aeronautico.xlsx')
sh4_codes = df_setor_aeronautico['CO_SH4']
display(sh4_codes.head())

0    8801
1    8801
2    8801
3    8802
4    8802
Name: CO_SH4, dtype: int64

## Processar arquivos de exportação

### Subtask:
Iterar sobre a lista de arquivos CSV de exportação.


**Reasoning**:
Create a list of the export file paths and iterate through them.



In [ ]:
export_files = ["/content/EXP_2021_MUN.csv", "/content/EXP_2022_MUN.csv", "/content/EXP_2023_MUN.csv", "/content/EXP_2024_MUN.csv", "/content/EXP_2025_MUN.csv"]

for file_path in export_files:
    print(f"Processing file: {file_path}")
    # The processing for each file will be done in subsequent steps.

Processing file: /content/EXP_2021_MUN.csv
Processing file: /content/EXP_2022_MUN.csv
Processing file: /content/EXP_2023_MUN.csv
Processing file: /content/EXP_2024_MUN.csv
Processing file: /content/EXP_2025_MUN.csv


## Carregar arquivo de exportação individual

### Subtask:
Carregar cada arquivo CSV de exportação em um DataFrame do pandas.


**Reasoning**:
Inside the loop, read the current CSV file into a pandas DataFrame, handling potential encoding and delimiter issues. Then display the head and columns of the loaded DataFrame to verify.



In [ ]:
export_files = ["/content/EXP_2021_MUN.csv", "/content/EXP_2022_MUN.csv", "/content/EXP_2023_MUN.csv", "/content/EXP_2024_MUN.csv", "/content/EXP_2025_MUN.csv"]

for file_path in export_files:
    print(f"Processing file: {file_path}")
    try:
        # Attempt to read with default parameters first
        df_export = pd.read_csv(file_path)
    except UnicodeDecodeError:
        # If UnicodeDecodeError occurs, try with 'latin1' encoding
        try:
            df_export = pd.read_csv(file_path, encoding='latin1')
        except:
            # If 'latin1' fails, try with 'ISO-8859-1'
            df_export = pd.read_csv(file_path, encoding='ISO-8859-1')
    except Exception as e:
        print(f"Error reading file {file_path}: {e}")
        continue

    # Check if the data is in a single column and if it's delimited by semicolon
    if df_export.shape[1] == 1 and df_export.columns[0].count(';') > 0:
        try:
            df_export = pd.read_csv(file_path, encoding='latin1', sep=';')
        except:
             df_export = pd.read_csv(file_path, encoding='ISO-8859-1', sep=';')

    print(f"Head of {file_path}:")
    display(df_export.head())
    print(f"Columns of {file_path}:")
    print(df_export.columns)

Processing file: /content/EXP_2021_MUN.csv
Head of /content/EXP_2021_MUN.csv:


CO_ANO  CO_MES   SH4  CO_PAIS SG_UF_MUN   CO_MUN  KG_LIQUIDO  VL_FOB
0    2021       3  3305      493        SP  3434401         966    7788
1    2021       7  8419      386        SP  3409502           1    1544
2    2021      12  8511       63        PR  4106902          73     907
3    2021       7  6204      586        SC  4202909          75    3942
4    2021       8  8708       63        SP  3440002        6949   29604

Columns of /content/EXP_2021_MUN.csv:
Index(['CO_ANO', 'CO_MES', 'SH4', 'CO_PAIS', 'SG_UF_MUN', 'CO_MUN',
       'KG_LIQUIDO', 'VL_FOB'],
      dtype='object')
Processing file: /content/EXP_2022_MUN.csv
Head of /content/EXP_2022_MUN.csv:


CO_ANO  CO_MES   SH4  CO_PAIS SG_UF_MUN   CO_MUN  KG_LIQUIDO    VL_FOB
0    2022       4  9403      845        PR  4101002       53482     47615
1    2022      10  8501      386        SC  4208906      220899   1348091
2    2022       9  1902      580        ES  3205309         564      2264
3    2022       1  6206       63        SC  4216602          85      6642
4    2022       2  8501      249        SC  4208906     2942192  12155328

Columns of /content/EXP_2022_MUN.csv:
Index(['CO_ANO', 'CO_MES', 'SH4', 'CO_PAIS', 'SG_UF_MUN', 'CO_MUN',
       'KG_LIQUIDO', 'VL_FOB'],
      dtype='object')
Processing file: /content/EXP_2023_MUN.csv
Head of /content/EXP_2023_MUN.csv:


CO_ANO  CO_MES   SH4  CO_PAIS SG_UF_MUN   CO_MUN  KG_LIQUIDO  VL_FOB
0    2023      10  2209      127        BA  2919553          36      53
1    2023      11  4009      493        SP  3456701         473   14008
2    2023      11  8536      365        SP  3438709           2     181
3    2023       8  8512       63        RJ  3304201       15627  239105
4    2023       6  1902      580        PR  4118204         287     368

Columns of /content/EXP_2023_MUN.csv:
Index(['CO_ANO', 'CO_MES', 'SH4', 'CO_PAIS', 'SG_UF_MUN', 'CO_MUN',
       'KG_LIQUIDO', 'VL_FOB'],
      dtype='object')
Processing file: /content/EXP_2024_MUN.csv
Head of /content/EXP_2024_MUN.csv:


CO_ANO  CO_MES   SH4  CO_PAIS SG_UF_MUN   CO_MUN  KG_LIQUIDO  VL_FOB
0    2024       2  8708      158        MG  3106705        3346   97359
1    2024       9  1516      845        SP  3425904      349090  541361
2    2024       7  8536      158        SC  4208906        1711   20706
3    2024       8  4010       63        SP  3452205        4586  176026
4    2024      12   701      476        SC  4216206         330     274

Columns of /content/EXP_2024_MUN.csv:
Index(['CO_ANO', 'CO_MES', 'SH4', 'CO_PAIS', 'SG_UF_MUN', 'CO_MUN',
       'KG_LIQUIDO', 'VL_FOB'],
      dtype='object')
Processing file: /content/EXP_2025_MUN.csv
Head of /content/EXP_2025_MUN.csv:


CO_ANO  CO_MES   SH4  CO_PAIS SG_UF_MUN   CO_MUN  KG_LIQUIDO  VL_FOB
0    2025       6   709      477        ES  3205002          95     273
1    2025       1   709      476        ES  3205309         545     836
2    2025       7  9403      249        SP  3420509        2439   13315
3    2025       1  8413       77        ES  3205309           1     364
4    2025       7  8483      493        SP  3448708        1303   22984

Columns of /content/EXP_2025_MUN.csv:
Index(['CO_ANO', 'CO_MES', 'SH4', 'CO_PAIS', 'SG_UF_MUN', 'CO_MUN',
       'KG_LIQUIDO', 'VL_FOB'],
      dtype='object')


**Reasoning**:
All the export files have been successfully loaded into DataFrames within the loop. The next step is to filter each of these DataFrames using the `sh4_codes` extracted earlier and then concatenate the filtered DataFrames into a single DataFrame.



In [ ]:
filtered_dfs = []

for file_path in export_files:
    print(f"Processing file: {file_path}")
    try:
        df_export = pd.read_csv(file_path, encoding='latin1', sep=';')
    except:
         df_export = pd.read_csv(file_path, encoding='ISO-8859-1', sep=';')

    df_exp_aeronautico_single_file = df_export[df_export['SH4'].isin(sh4_codes)]
    filtered_dfs.append(df_exp_aeronautico_single_file)

df_exp_aeronautico = pd.concat(filtered_dfs, ignore_index=True)
display(df_exp_aeronautico.head())
print(df_exp_aeronautico.columns)

Processing file: /content/EXP_2021_MUN.csv
Processing file: /content/EXP_2022_MUN.csv
Processing file: /content/EXP_2023_MUN.csv
Processing file: /content/EXP_2024_MUN.csv
Processing file: /content/EXP_2025_MUN.csv


CO_ANO  CO_MES   SH4  CO_PAIS SG_UF_MUN   CO_MUN  KG_LIQUIDO    VL_FOB
0    2021       9  8802      249        SP  3416853        9973  14903100
1    2021       8  8802       69        SP  3425904        2245    480000
2    2021       1  8802      589        AC  1200252          90      2444
3    2021       7  8805      249        SP  3418800           7      5300
4    2021       9  8802      767        SP  3443402           1      5966

Index(['CO_ANO', 'CO_MES', 'SH4', 'CO_PAIS', 'SG_UF_MUN', 'CO_MUN',
       'KG_LIQUIDO', 'VL_FOB'],
      dtype='object')


## Salvar dados filtrados

### Subtask:
Salvar o DataFrame filtrado com os dados de exportação do setor aeronáutico (`df_exp_aeronautico`) em um novo arquivo Excel.


**Reasoning**:
Save the filtered DataFrame `df_exp_aeronautico` to an Excel file named 'EXP_Aeronautico_consolidado.xlsx' without including the index.



In [ ]:
df_exp_aeronautico.to_excel('EXP_Aeronautico_consolidado.xlsx', index=False)

## Summary:

### Data Analysis Key Findings

*   The SH4 codes from the aeronautical sector file were successfully extracted and used as a filter.
*   Multiple export data files from 2021 to 2025 were successfully loaded, handling potential encoding and delimiter issues (semicolon).
*   The export data from all processed CSV files was filtered to include only entries matching the specified SH4 codes.
*   The filtered data from all years was concatenated into a single DataFrame.
*   The final consolidated and filtered export data for the aeronautical sector was saved to a new Excel file named `EXP_Aeronautico_consolidado.xlsx`.

### Insights or Next Steps

*   The consolidated data in `EXP_Aeronautico_consolidado.xlsx` can now be used for further analysis, such as identifying trends in aeronautical exports over the years 2021-2025, determining key export destinations, or analyzing the value of these exports.
*   Consider adding error handling to log files that fail to load or process correctly during the iteration.


# Task
Analise todos os dados dos arquivos: "/content/EXP_2021_MUN.csv", "/content/EXP_2022_MUN.csv", "/content/EXP_2023_MUN.csv", "/content/EXP_2024_MUN.csv", "/content/EXP_2025_MUN.csv", e faça um filtro separando somente os dados que forem iguais aos seguintes dados: CO_SH4 do arquivo "/content/setor_aeronautico.xlsx"; NO_NCM_POR do arquivo "/content/NCMsetor_aeronautico.xlsx". Gere uma planilha em excel com esses dados.

## Carregar dados de referência

### Subtask:
Carregar os arquivos `/content/setor_aeronautico.xlsx` e `/content/NCMsetor_aeronautico.xlsx` em DataFrames do pandas e extrair as colunas relevantes (`CO_SH4` e `NO_NCM_POR`).


**Reasoning**:
Load the two Excel files into pandas DataFrames and extract the specified columns. Then display the heads of the extracted columns to verify.



In [ ]:
df_setor_aeronautico = pd.read_excel('/content/setor_aeronautico.xlsx')
sh4_codes = df_setor_aeronautico['CO_SH4']

df_ncm_aeronautico = pd.read_excel('/content/NCMsetor_aeronautico.xlsx')
ncm_descriptions = df_ncm_aeronautico['NO_NCM_POR']

display(sh4_codes.head())
display(ncm_descriptions.head())

0    8801
1    8801
2    8801
3    8802
4    8802
Name: CO_SH4, dtype: int64

0    Outros veículos aéreos (aeronaves) não tripula...
1    Veículos aéreos (aeronaves) não tripulados, co...
2    Consumo de bordo - combustíveis e lubrificante...
3                    Motores de explosão, para aviação
4    Limpadores de pára-brisas elétricos, para aero...
Name: NO_NCM_POR, dtype: object

## Processar arquivos de exportação

### Subtask:
Iterar sobre a lista de arquivos CSV de exportação.


**Reasoning**:
Define the list of export file paths and iterate through it, printing the name of the file being processed in each iteration.



In [ ]:
export_files = ["/content/EXP_2021_MUN.csv", "/content/EXP_2022_MUN.csv", "/content/EXP_2023_MUN.csv", "/content/EXP_2024_MUN.csv", "/content/EXP_2025_MUN.csv"]

for file_path in export_files:
    print(f"Processing file: {file_path}")

Processing file: /content/EXP_2021_MUN.csv
Processing file: /content/EXP_2022_MUN.csv
Processing file: /content/EXP_2023_MUN.csv
Processing file: /content/EXP_2024_MUN.csv
Processing file: /content/EXP_2025_MUN.csv


**Reasoning**:
Inside the loop, load each CSV file into a DataFrame, handling potential encoding and delimiter issues, and display the head and columns to verify the loading process.



In [ ]:
export_files = ["/content/EXP_2021_MUN.csv", "/content/EXP_2022_MUN.csv", "/content/EXP_2023_MUN.csv", "/content/EXP_2024_MUN.csv", "/content/EXP_2025_MUN.csv"]

for file_path in export_files:
    print(f"Processing file: {file_path}")
    try:
        df_export = pd.read_csv(file_path)
    except UnicodeDecodeError:
        try:
            df_export = pd.read_csv(file_path, encoding='latin1')
        except:
            df_export = pd.read_csv(file_path, encoding='ISO-8859-1')
    except Exception as e:
        print(f"Error reading file {file_path}: {e}")
        continue

    if df_export.shape[1] == 1 and df_export.columns[0].count(';') > 0:
        try:
            df_export = pd.read_csv(file_path, encoding='latin1', sep=';')
        except:
             df_export = pd.read_csv(file_path, encoding='ISO-8859-1', sep=';')

    print(f"Head of {file_path}:")
    display(df_export.head())
    print(f"Columns of {file_path}:")
    print(df_export.columns)

Processing file: /content/EXP_2021_MUN.csv
Head of /content/EXP_2021_MUN.csv:


CO_ANO  CO_MES   SH4  CO_PAIS SG_UF_MUN   CO_MUN  KG_LIQUIDO  VL_FOB
0    2021       3  3305      493        SP  3434401         966    7788
1    2021       7  8419      386        SP  3409502           1    1544
2    2021      12  8511       63        PR  4106902          73     907
3    2021       7  6204      586        SC  4202909          75    3942
4    2021       8  8708       63        SP  3440002        6949   29604

Columns of /content/EXP_2021_MUN.csv:
Index(['CO_ANO', 'CO_MES', 'SH4', 'CO_PAIS', 'SG_UF_MUN', 'CO_MUN',
       'KG_LIQUIDO', 'VL_FOB'],
      dtype='object')
Processing file: /content/EXP_2022_MUN.csv
Head of /content/EXP_2022_MUN.csv:


CO_ANO  CO_MES   SH4  CO_PAIS SG_UF_MUN   CO_MUN  KG_LIQUIDO    VL_FOB
0    2022       4  9403      845        PR  4101002       53482     47615
1    2022      10  8501      386        SC  4208906      220899   1348091
2    2022       9  1902      580        ES  3205309         564      2264
3    2022       1  6206       63        SC  4216602          85      6642
4    2022       2  8501      249        SC  4208906     2942192  12155328

Columns of /content/EXP_2022_MUN.csv:
Index(['CO_ANO', 'CO_MES', 'SH4', 'CO_PAIS', 'SG_UF_MUN', 'CO_MUN',
       'KG_LIQUIDO', 'VL_FOB'],
      dtype='object')
Processing file: /content/EXP_2023_MUN.csv
Head of /content/EXP_2023_MUN.csv:


CO_ANO  CO_MES   SH4  CO_PAIS SG_UF_MUN   CO_MUN  KG_LIQUIDO  VL_FOB
0    2023      10  2209      127        BA  2919553          36      53
1    2023      11  4009      493        SP  3456701         473   14008
2    2023      11  8536      365        SP  3438709           2     181
3    2023       8  8512       63        RJ  3304201       15627  239105
4    2023       6  1902      580        PR  4118204         287     368

Columns of /content/EXP_2023_MUN.csv:
Index(['CO_ANO', 'CO_MES', 'SH4', 'CO_PAIS', 'SG_UF_MUN', 'CO_MUN',
       'KG_LIQUIDO', 'VL_FOB'],
      dtype='object')
Processing file: /content/EXP_2024_MUN.csv
Head of /content/EXP_2024_MUN.csv:


CO_ANO  CO_MES   SH4  CO_PAIS SG_UF_MUN   CO_MUN  KG_LIQUIDO  VL_FOB
0    2024       2  8708      158        MG  3106705        3346   97359
1    2024       9  1516      845        SP  3425904      349090  541361
2    2024       7  8536      158        SC  4208906        1711   20706
3    2024       8  4010       63        SP  3452205        4586  176026
4    2024      12   701      476        SC  4216206         330     274

Columns of /content/EXP_2024_MUN.csv:
Index(['CO_ANO', 'CO_MES', 'SH4', 'CO_PAIS', 'SG_UF_MUN', 'CO_MUN',
       'KG_LIQUIDO', 'VL_FOB'],
      dtype='object')
Processing file: /content/EXP_2025_MUN.csv
Head of /content/EXP_2025_MUN.csv:


CO_ANO  CO_MES   SH4  CO_PAIS SG_UF_MUN   CO_MUN  KG_LIQUIDO  VL_FOB
0    2025       6   709      477        ES  3205002          95     273
1    2025       1   709      476        ES  3205309         545     836
2    2025       7  9403      249        SP  3420509        2439   13315
3    2025       1  8413       77        ES  3205309           1     364
4    2025       7  8483      493        SP  3448708        1303   22984

Columns of /content/EXP_2025_MUN.csv:
Index(['CO_ANO', 'CO_MES', 'SH4', 'CO_PAIS', 'SG_UF_MUN', 'CO_MUN',
       'KG_LIQUIDO', 'VL_FOB'],
      dtype='object')


**Reasoning**:
All files have been successfully loaded and their structure verified. The next step is to filter each DataFrame based on the `sh4_codes` and `ncm_descriptions` and then concatenate the filtered DataFrames into a single DataFrame.



In [ ]:
filtered_dfs = []

for file_path in export_files:
    print(f"Processing file: {file_path}")
    try:
        df_export = pd.read_csv(file_path, encoding='latin1', sep=';')
    except:
         df_export = pd.read_csv(file_path, encoding='ISO-8859-1', sep=';')

    df_exp_aeronautico_single_file = df_export[
        df_export['SH4'].isin(sh4_codes) |
        df_export['NO_NCM_POR'].isin(ncm_descriptions)
    ]
    filtered_dfs.append(df_exp_aeronautico_single_file)

df_exp_aeronautico = pd.concat(filtered_dfs, ignore_index=True)
display(df_exp_aeronautico.head())
print(df_exp_aeronautico.columns)

Processing file: /content/EXP_2021_MUN.csv


KeyError: 'NO_NCM_POR'

**Reasoning**:
The filtering failed because the column 'NO_NCM_POR' does not exist in the export data DataFrames. I need to check the columns of one of the export DataFrames again to confirm the available column names for filtering.



In [ ]:
print(df_export.columns)

Index(['CO_ANO', 'CO_MES', 'SH4', 'CO_PAIS', 'SG_UF_MUN', 'CO_MUN',
       'KG_LIQUIDO', 'VL_FOB'],
      dtype='object')


**Reasoning**:
The column names of the export data were printed, confirming that 'NO_NCM_POR' is not present. The filtering should only be based on the 'SH4' column. I will retry the filtering and concatenation using only the 'SH4' column.



In [ ]:
filtered_dfs = []

for file_path in export_files:
    print(f"Processing file: {file_path}")
    try:
        df_export = pd.read_csv(file_path, encoding='latin1', sep=';')
    except:
         df_export = pd.read_csv(file_path, encoding='ISO-8859-1', sep=';')

    df_exp_aeronautico_single_file = df_export[df_export['SH4'].isin(sh4_codes)]
    filtered_dfs.append(df_exp_aeronautico_single_file)

df_exp_aeronautico = pd.concat(filtered_dfs, ignore_index=True)
display(df_exp_aeronautico.head())
print(df_exp_aeronautico.columns)

Processing file: /content/EXP_2021_MUN.csv
Processing file: /content/EXP_2022_MUN.csv
Processing file: /content/EXP_2023_MUN.csv
Processing file: /content/EXP_2024_MUN.csv
Processing file: /content/EXP_2025_MUN.csv


CO_ANO  CO_MES   SH4  CO_PAIS SG_UF_MUN   CO_MUN  KG_LIQUIDO    VL_FOB
0    2021       9  8802      249        SP  3416853        9973  14903100
1    2021       8  8802       69        SP  3425904        2245    480000
2    2021       1  8802      589        AC  1200252          90      2444
3    2021       7  8805      249        SP  3418800           7      5300
4    2021       9  8802      767        SP  3443402           1      5966

Index(['CO_ANO', 'CO_MES', 'SH4', 'CO_PAIS', 'SG_UF_MUN', 'CO_MUN',
       'KG_LIQUIDO', 'VL_FOB'],
      dtype='object')


## Filtrar dados de exportação

### Subtask:
Filtrar o DataFrame de exportação atual para incluir apenas as linhas onde o `SH4` corresponde aos valores `CO_SH4` do arquivo do setor aeronáutico.


**Reasoning**:
Filter the export data DataFrame to include only the rows where the 'SH4' column values are present in the 'sh4_codes' Series, and also filter for rows where 'NO_NCM_POR' matches the descriptions in the 'ncm_descriptions' Series. Given the previous error, I will first ensure that 'NO_NCM_POR' column exists before attempting to filter.



In [ ]:
filtered_dfs = []

for file_path in export_files:
    print(f"Processing file: {file_path}")
    try:
        df_export = pd.read_csv(file_path, encoding='latin1', sep=';')
    except:
         df_export = pd.read_csv(file_path, encoding='ISO-8859-1', sep=';')

    if 'NO_NCM_POR' in df_export.columns:
        df_exp_aeronautico_single_file = df_export[
            df_export['SH4'].isin(sh4_codes) &
            df_export['NO_NCM_POR'].isin(ncm_descriptions)
        ]
    else:
        df_exp_aeronautico_single_file = df_export[df_export['SH4'].isin(sh4_codes)]


    filtered_dfs.append(df_exp_aeronautico_single_file)

df_exp_aeronautico = pd.concat(filtered_dfs, ignore_index=True)
display(df_exp_aeronautico.head())
print(df_exp_aeronautico.columns)

Processing file: /content/EXP_2021_MUN.csv
Processing file: /content/EXP_2022_MUN.csv
Processing file: /content/EXP_2023_MUN.csv
Processing file: /content/EXP_2024_MUN.csv
Processing file: /content/EXP_2025_MUN.csv


CO_ANO  CO_MES   SH4  CO_PAIS SG_UF_MUN   CO_MUN  KG_LIQUIDO    VL_FOB
0    2021       9  8802      249        SP  3416853        9973  14903100
1    2021       8  8802       69        SP  3425904        2245    480000
2    2021       1  8802      589        AC  1200252          90      2444
3    2021       7  8805      249        SP  3418800           7      5300
4    2021       9  8802      767        SP  3443402           1      5966

Index(['CO_ANO', 'CO_MES', 'SH4', 'CO_PAIS', 'SG_UF_MUN', 'CO_MUN',
       'KG_LIQUIDO', 'VL_FOB'],
      dtype='object')


## Salvar dados filtrados consolidados

### Subtask:
Salvar o DataFrame consolidado e filtrado em um novo arquivo Excel com um nome indicativo (por exemplo, `EXP_Aeronautico_NCM_consolidado.xlsx`).


**Reasoning**:
Save the filtered and consolidated DataFrame to an Excel file named 'EXP_Aeronautico_NCM_consolidado.xlsx' without including the index.



In [ ]:
df_exp_aeronautico.to_excel('EXP_Aeronautico_NCM_consolidado.xlsx', index=False)

## Summary:

### Data Analysis Key Findings

*   Export data from 2021 to 2025 across multiple municipalities was successfully loaded and consolidated.
*   Aeronautical sector codes (SH4) and NCM descriptions were loaded from reference files.
*   The consolidated export data was filtered based on the provided aeronautical sector SH4 codes.
*   Filtering by NCM descriptions was skipped due to the absence of the 'NO\_NCM\_POR' column in the export data files.
*   The filtered data was successfully saved to a new Excel file named `EXP_Aeronautico_NCM_consolidado.xlsx`.

### Insights or Next Steps

*   Investigate if there is an alternative column in the export data that can be used for filtering by NCM description if required for further analysis.
*   Analyze the content of the `EXP_Aeronautico_NCM_consolidado.xlsx` file to identify trends or patterns in aeronautical sector exports by municipality and year.


# Task
Analise o arquivo "/content/EXP_Aeronautico_NCM_consolidado.xlsx", elimine da coluna SG_UF_MUN todos dados que não são SP. Depois faça o cruzamento dos dados desse arquivo com os arquivos: "/content/PAIS.csv", "/content/UF_MUN.csv", "/content/UF_MUN_SP.xlsx" substituindo os códigos pelos nomes, conforme os arquivos. Gere uma Planilha em excel.

## Carregar os dados consolidados

### Subtask:
Carregar o arquivo `/content/EXP_Aeronautico_NCM_consolidado.xlsx` em um DataFrame do pandas.


**Reasoning**:
Import the pandas library and load the Excel file into a DataFrame. Then display the head of the DataFrame and its columns to verify the data loading.



In [ ]:
import pandas as pd

df_exp_aeronautico_consolidado = pd.read_excel('/content/EXP_Aeronautico_NCM_consolidado.xlsx')
display(df_exp_aeronautico_consolidado.head())
print(df_exp_aeronautico_consolidado.columns)

CO_ANO  CO_MES   SH4  CO_PAIS SG_UF_MUN   CO_MUN  KG_LIQUIDO    VL_FOB
0    2021       9  8802      249        SP  3416853        9973  14903100
1    2021       8  8802       69        SP  3425904        2245    480000
2    2021       1  8802      589        AC  1200252          90      2444
3    2021       7  8805      249        SP  3418800           7      5300
4    2021       9  8802      767        SP  3443402           1      5966

Index(['CO_ANO', 'CO_MES', 'SH4', 'CO_PAIS', 'SG_UF_MUN', 'CO_MUN',
       'KG_LIQUIDO', 'VL_FOB'],
      dtype='object')


## Filtrar por estado

### Subtask:
Filtrar o DataFrame consolidado para incluir apenas as linhas onde a coluna `SG_UF_MUN` é igual a 'SP'.


**Reasoning**:
Filter the DataFrame to include only rows where 'SG_UF_MUN' is 'SP' and display the head of the filtered DataFrame.



In [ ]:
df_exp_aeronautico_sp = df_exp_aeronautico_consolidado[df_exp_aeronautico_consolidado['SG_UF_MUN'] == 'SP']
display(df_exp_aeronautico_sp.head())

CO_ANO  CO_MES   SH4  CO_PAIS SG_UF_MUN   CO_MUN  KG_LIQUIDO    VL_FOB
0    2021       9  8802      249        SP  3416853        9973  14903100
1    2021       8  8802       69        SP  3425904        2245    480000
3    2021       7  8805      249        SP  3418800           7      5300
4    2021       9  8802      767        SP  3443402           1      5966
5    2021      11  8802      573        SP  3449904       33686  33229253

## Carregar dados de referência para cruzamento

### Subtask:
Carregar os arquivos `/content/PAIS.csv`, `/content/UF_MUN.csv` e `/content/UF_MUN_SP.xlsx` em DataFrames do pandas.


**Reasoning**:
Load the specified CSV and Excel files into pandas DataFrames and display their heads and columns to verify the loading.



In [ ]:
df_paises = pd.read_csv('/content/PAIS.csv')
display(df_paises.head())
print(df_paises.columns)

df_uf_mun = pd.read_csv('/content/UF_MUN.csv', encoding='latin1', sep=';')
display(df_uf_mun.head())
print(df_uf_mun.columns)

df_uf_mun_sp_ref = pd.read_excel('/content/UF_MUN_SP.xlsx')
display(df_uf_mun_sp_ref.head())
print(df_uf_mun_sp_ref.columns)

UnicodeDecodeError: 'utf-8' codec can't decode byte 0xe3 in position 101: invalid continuation byte

**Reasoning**:
The previous attempt to read '/content/PAIS.csv' failed due to a UnicodeDecodeError. I will try reading this file again, specifying 'latin1' encoding, which is a common alternative for files that are not UTF-8.



In [ ]:
df_paises = pd.read_csv('/content/PAIS.csv', encoding='latin1')
display(df_paises.head())
print(df_paises.columns)

df_uf_mun = pd.read_csv('/content/UF_MUN.csv', encoding='latin1', sep=';')
display(df_uf_mun.head())
print(df_uf_mun.columns)

df_uf_mun_sp_ref = pd.read_excel('/content/UF_MUN_SP.xlsx')
display(df_uf_mun_sp_ref.head())
print(df_uf_mun_sp_ref.columns)

ParserError: Error tokenizing data. C error: Expected 1 fields in line 4, saw 3


**Reasoning**:
The previous attempt to read '/content/PAIS.csv' with 'latin1' encoding resulted in a ParserError, indicating a delimiter issue. I will try reading this file again, keeping the 'latin1' encoding and explicitly setting the delimiter to a semicolon, which is a common delimiter in CSV-like files.



In [ ]:
df_paises = pd.read_csv('/content/PAIS.csv', encoding='latin1', sep=';')
display(df_paises.head())
print(df_paises.columns)

df_uf_mun = pd.read_csv('/content/UF_MUN.csv', encoding='latin1', sep=';')
display(df_uf_mun.head())
print(df_uf_mun.columns)

df_uf_mun_sp_ref = pd.read_excel('/content/UF_MUN_SP.xlsx')
display(df_uf_mun_sp_ref.head())
print(df_uf_mun_sp_ref.columns)

CO_PAIS  CO_PAIS_ISON3 CO_PAIS_ISOA3                 NO_PAIS  \
0        0            898           ZZZ            Não Definido   
1       13              4           AFG             Afeganistão   
2       15            248           ALA            Aland, Ilhas   
3       17              8           ALB                 Albânia   
4       20            724           ESP  Alboran-Perejil, Ilhas   

                NO_PAIS_ING             NO_PAIS_ESP  
0               Not defined             No definido  
1               Afghanistan              Afganistan  
2             Aland Islands            Alans, Islas  
3                   Albania                 Albania  
4  Alboran-Perejil, Islands  Alboran-Perejil, Islas

Index(['CO_PAIS', 'CO_PAIS_ISON3', 'CO_PAIS_ISOA3', 'NO_PAIS', 'NO_PAIS_ING',
       'NO_PAIS_ESP'],
      dtype='object')


CO_MUN_GEO                  NO_MUN              NO_MUN_MIN SG_UF
0     2502003  BELEM DO BREJO DO CRUZ  Belém do Brejo do Cruz    PB
1     2502052      BERNARDINO BATISTA      Bernardino Batista    PB
2     2502102             BOA VENTURA             Boa Ventura    PB
3     2502151               BOA VISTA               Boa Vista    PB
4     2502201               BOM JESUS               Bom Jesus    PB

Index(['CO_MUN_GEO', 'NO_MUN', 'NO_MUN_MIN', 'SG_UF'], dtype='object')


CO_MUN_GEO                  NO_MUN              NO_MUN_MIN SG_UF
0     3446900             SANTA LUCIA             Santa Lúcia    SP
1     3447007    SANTA MARIA DA SERRA    Santa Maria da Serra    SP
2     3447106          SANTA MERCEDES          Santa Mercedes    SP
3     3447205  SANTANA DA PONTE PENSA  Santana da Ponte Pensa    SP
4     3447304     SANTANA DE PARNAIBA     Santana de Parnaíba    SP

Index(['CO_MUN_GEO', 'NO_MUN', 'NO_MUN_MIN', 'SG_UF'], dtype='object')


# Task
Analise o arquivo "/content/EXP_Aeronautico_SP_com_Nomes.xlsx", cruze as informações com o arquivo "/content/MUNICIPIOS.csv" para incluir a região do município, e gere uma planilha em excel com esses dados, separando por ano em abas.

## Carregar os dados de exportação com nomes

### Subtask:
Carregar o arquivo `/content/EXP_Aeronautico_SP_com_Nomes.xlsx` em um DataFrame do pandas.


**Reasoning**:
Import the pandas library and load the Excel file into a DataFrame. Then display the head of the DataFrame to verify the data loading.



In [ ]:
df_exp_aeronautico_sp_nomes = pd.read_excel('/content/EXP_Aeronautico_SP_com_Nomes.xlsx')
display(df_exp_aeronautico_sp_nomes.head())

CO_ANO  CO_MES   SH4  CO_PAIS SG_UF_MUN   CO_MUN  KG_LIQUIDO    VL_FOB  \
0    2021       9  8802      249        SP  3416853        9973  14903100   
1    2021       8  8802       69        SP  3425904        2245    480000   
2    2021       7  8805      249        SP  3418800           7      5300   
3    2021       9  8802      767        SP  3443402           1      5966   
4    2021      11  8802      573        SP  3449904       33686  33229253   

                   NO_PAIS               NO_MUN  
0           Estados Unidos       GAVIAO PEIXOTO  
1                Austrália              JUNDIAI  
2           Estados Unidos            GUARULHOS  
3                    Suíça       RIBEIRAO PRETO  
4  Países Baixos (Holanda)  SAO JOSE DOS CAMPOS

## Carregar os dados de municípios com região

### Subtask:
Carregar o arquivo `/content/MUNICIPIOS.csv` em um DataFrame do pandas e selecionar as colunas relevantes para o cruzamento (código do município e nome da região).


**Reasoning**:
Load the CSV file into a DataFrame, then display the head and columns to inspect the data and identify the relevant columns for municipality code and region name.



In [ ]:
df_municipios = pd.read_csv('/content/MUNICIPIOS.csv')
display(df_municipios.head())
print(df_municipios.columns)

UnicodeDecodeError: 'utf-8' codec can't decode byte 0xe3 in position 4: invalid continuation byte

**Reasoning**:
The previous attempt to read the CSV failed due to a UnicodeDecodeError. This suggests that the file is not encoded in UTF-8. I will try reading the file again, specifying 'latin1' encoding, which is a common alternative for files that are not UTF-8.



In [ ]:
df_municipios = pd.read_csv('/content/MUNICIPIOS.csv', encoding='latin1')
display(df_municipios.head())
print(df_municipios.columns)

Região           Município
0  Vale do Paraíba e Litoral Norte           Aparecida
1  Vale do Paraíba e Litoral Norte              Areias
2  Vale do Paraíba e Litoral Norte             Bananal
3  Vale do Paraíba e Litoral Norte  Cachoeira Paulista
4  Vale do Paraíba e Litoral Norte               Canas

Index(['Região', 'Município'], dtype='object')


## Realizar cruzamento de dados (região)

### Subtask:
Cruzar o DataFrame de exportação com nomes com o DataFrame de municípios com região para adicionar a informação da região.


**Reasoning**:
Merge the export data DataFrame with the municipalities DataFrame to add the region information based on the municipality name. Display the head of the resulting DataFrame.



In [ ]:
df_exp_aeronautico_sp_com_regiao = pd.merge(
    df_exp_aeronautico_sp_nomes,
    df_municipios[['Região', 'Município']],
    left_on='NO_MUN',
    right_on='Município',
    how='left'
)

# Drop the redundant 'Município' column after merging
df_exp_aeronautico_sp_com_regiao = df_exp_aeronautico_sp_com_regiao.drop('Município', axis=1)

display(df_exp_aeronautico_sp_com_regiao.head())

CO_ANO  CO_MES   SH4  CO_PAIS SG_UF_MUN   CO_MUN  KG_LIQUIDO    VL_FOB  \
0    2021       9  8802      249        SP  3416853        9973  14903100   
1    2021       8  8802       69        SP  3425904        2245    480000   
2    2021       7  8805      249        SP  3418800           7      5300   
3    2021       9  8802      767        SP  3443402           1      5966   
4    2021      11  8802      573        SP  3449904       33686  33229253   

                   NO_PAIS               NO_MUN Região  
0           Estados Unidos       GAVIAO PEIXOTO    NaN  
1                Austrália              JUNDIAI    NaN  
2           Estados Unidos            GUARULHOS    NaN  
3                    Suíça       RIBEIRAO PRETO    NaN  
4  Países Baixos (Holanda)  SAO JOSE DOS CAMPOS    NaN

**Reasoning**:
The previous merge resulted in many NaN values in the 'Região' column, indicating that the municipality names in the two DataFrames do not match exactly. I need to inspect the unique municipality names in both DataFrames to understand the discrepancies and potentially clean or standardize them before merging.



In [ ]:
print("Unique municipalities in df_exp_aeronautico_sp_nomes:")
print(df_exp_aeronautico_sp_nomes['NO_MUN'].unique())

print("\nUnique municipalities in df_municipios:")
print(df_municipios['Município'].unique())

Unique municipalities in df_exp_aeronautico_sp_nomes:
['GAVIAO PEIXOTO' 'JUNDIAI' 'GUARULHOS' 'RIBEIRAO PRETO'
 'SAO JOSE DOS CAMPOS' 'IPEUNA' 'OSASCO' 'SAO PAULO' 'LORENA'
 'SAO JOAO DA BOA VISTA' 'CAMPINAS' 'JAGUARIUNA' 'SANTANA DE PARNAIBA'
 'SOROCABA' 'PRESIDENTE PRUDENTE' 'LEME' 'TAUBATE' 'SANTOS' 'SUMARE'
 'SAO JOSE DO RIO PRETO' 'COTIA' 'BARUERI' 'SAO CARLOS' 'BOITUVA'
 'SAO CAETANO DO SUL' 'SAO JOAQUIM DA BARRA' 'ALFREDO MARCONDES'
 'ARACATUBA' 'PIRACICABA' 'LINS' 'JACAREI' 'BRAGANCA PAULISTA' 'VINHEDO'
 'SERTAOZINHO']

Unique municipalities in df_municipios:
['Aparecida' 'Areias' 'Bananal' 'Cachoeira Paulista' 'Canas'
 'Caraguatatuba' 'Cruzeiro' 'Guaratinguetá' 'Ilhabela' 'Jacareí'
 'Jambeiro' 'Lorena' 'Natividade da Serra' 'Pindamonhangaba' 'Potim'
 'Queluz' 'Roseira' 'Santa Branca' 'São Bento do Sapucaí'
 'São José do Barreiro' 'São José dos Campos' 'São Luiz do Paraitinga'
 'São Sebastião' 'Tremembé' 'Ubatuba' 'Taubaté' 'Bertioga' 'Cubatão'
 'Guarujá' 'Itanhaém' 'Mongaguá' 

**Reasoning**:
The municipality names have different casing and some have accented characters in `df_municipios` which might be causing the merge to fail. I will standardize the municipality names in both dataframes by converting them to uppercase and removing accented characters before attempting the merge again.



In [ ]:
import unicodedata

def remove_accents(input_str):
    if isinstance(input_str, str):
        nfkd_form = unicodedata.normalize('NFKD', input_str)
        return "".join([c for c in nfkd_form if not unicodedata.combining(c)])
    return input_str

df_exp_aeronautico_sp_nomes['NO_MUN_cleaned'] = df_exp_aeronautico_sp_nomes['NO_MUN'].apply(lambda x: remove_accents(x).upper())
df_municipios['Município_cleaned'] = df_municipios['Município'].apply(lambda x: remove_accents(x).upper())

df_exp_aeronautico_sp_com_regiao = pd.merge(
    df_exp_aeronautico_sp_nomes,
    df_municipios[['Região', 'Município_cleaned']],
    left_on='NO_MUN_cleaned',
    right_on='Município_cleaned',
    how='left'
)

# Drop the redundant cleaned columns and the original 'Município' column from df_municipios
df_exp_aeronautico_sp_com_regiao = df_exp_aeronautico_sp_com_regiao.drop(['NO_MUN_cleaned', 'Município_cleaned'], axis=1)

display(df_exp_aeronautico_sp_com_regiao.head())

CO_ANO  CO_MES   SH4  CO_PAIS SG_UF_MUN   CO_MUN  KG_LIQUIDO    VL_FOB  \
0    2021       9  8802      249        SP  3416853        9973  14903100   
1    2021       8  8802       69        SP  3425904        2245    480000   
2    2021       7  8805      249        SP  3418800           7      5300   
3    2021       9  8802      767        SP  3443402           1      5966   
4    2021      11  8802      573        SP  3449904       33686  33229253   

                   NO_PAIS               NO_MUN  \
0           Estados Unidos       GAVIAO PEIXOTO   
1                Austrália              JUNDIAI   
2           Estados Unidos            GUARULHOS   
3                    Suíça       RIBEIRAO PRETO   
4  Países Baixos (Holanda)  SAO JOSE DOS CAMPOS   

                            Região  
0                              NaN  
1                          Jundiaí  
2                        São Paulo  
3                              NaN  
4  Vale do Paraíba e Litoral Norte

**Reasoning**:
The municipality names have been cleaned and the merge was performed. However, there are still NaN values in the 'Região' column. This indicates that some municipalities in `df_exp_aeronautico_sp_nomes` were not found in `df_municipios` even after cleaning. I need to identify which municipalities from the export data are missing from the municipalities reference data.



In [ ]:
missing_municipalities = df_exp_aeronautico_sp_com_regiao[df_exp_aeronautico_sp_com_regiao['Região'].isna()]['NO_MUN'].unique()
print("Municipalities in df_exp_aeronautico_sp_nomes not found in df_municipios:")
print(missing_municipalities)

Municipalities in df_exp_aeronautico_sp_nomes not found in df_municipios:
['GAVIAO PEIXOTO' 'RIBEIRAO PRETO' 'IPEUNA' 'SAO JOAO DA BOA VISTA'
 'JAGUARIUNA' 'SANTANA DE PARNAIBA' 'PRESIDENTE PRUDENTE' 'LEME'
 'SAO CARLOS' 'SAO JOAQUIM DA BARRA' 'ALFREDO MARCONDES' 'ARACATUBA'
 'LINS' 'BRAGANCA PAULISTA' 'SERTAOZINHO']


**Reasoning**:
The missing municipalities have been identified. I need to manually map these missing municipalities to their respective regions based on external knowledge or a more comprehensive reference file if available. Since a more comprehensive file is not provided, I will create a dictionary with the missing municipalities and their regions and then use this dictionary to fill the missing 'Região' values in the DataFrame.



In [ ]:
missing_municipalities_mapping = {
    'GAVIAO PEIXOTO': 'Araraquara',
    'RIBEIRAO PRETO': 'Ribeirão Preto',
    'IPEUNA': 'Piracicaba',
    'SAO JOAO DA BOA VISTA': 'São João da Boa Vista',
    'JAGUARIUNA': 'Campinas',
    'SANTANA DE PARNAIBA': 'São Paulo',
    'PRESIDENTE PRUDENTE': 'Presidente Prudente',
    'LEME': 'Campinas',
    'SAO CARLOS': 'São Carlos',
    'SAO JOAQUIM DA BARRA': 'Franca',
    'ALFREDO MARCONDES': 'Presidente Prudente',
    'ARACATUBA': 'Araçatuba',
    'LINS': 'Bauru',
    'BRAGANCA PAULISTA': 'Campinas',
    'SERTAOZINHO': 'Ribeirão Preto'
}

df_exp_aeronautico_sp_com_regiao['Região'] = df_exp_aeronautico_sp_com_regiao['NO_MUN'].apply(
    lambda x: missing_municipalities_mapping.get(x, df_exp_aeronautico_sp_com_regiao[df_exp_aeronautico_sp_com_regiao['NO_MUN'] == x]['Região'].iloc[0])
)

display(df_exp_aeronautico_sp_com_regiao.head())

CO_ANO  CO_MES   SH4  CO_PAIS SG_UF_MUN   CO_MUN  KG_LIQUIDO    VL_FOB  \
0    2021       9  8802      249        SP  3416853        9973  14903100   
1    2021       8  8802       69        SP  3425904        2245    480000   
2    2021       7  8805      249        SP  3418800           7      5300   
3    2021       9  8802      767        SP  3443402           1      5966   
4    2021      11  8802      573        SP  3449904       33686  33229253   

                   NO_PAIS               NO_MUN  \
0           Estados Unidos       GAVIAO PEIXOTO   
1                Austrália              JUNDIAI   
2           Estados Unidos            GUARULHOS   
3                    Suíça       RIBEIRAO PRETO   
4  Países Baixos (Holanda)  SAO JOSE DOS CAMPOS   

                            Região  
0                       Araraquara  
1                          Jundiaí  
2                        São Paulo  
3                   Ribeirão Preto  
4  Vale do Paraíba e Litoral Norte

## Salvar em excel com abas por ano

### Subtask:
Salvar o DataFrame resultante em um novo arquivo Excel, com cada ano em uma aba separada.


**Reasoning**:
Create an Excel writer object and iterate through the unique years in the DataFrame to save each year's data to a separate sheet.



In [ ]:
excel_file_name = 'EXP_Aeronautico_SP_com_Regiao_por_Ano.xlsx'
with pd.ExcelWriter(excel_file_name) as writer:
    for year in df_exp_aeronautico_sp_com_regiao['CO_ANO'].unique():
        df_year = df_exp_aeronautico_sp_com_regiao[df_exp_aeronautico_sp_com_regiao['CO_ANO'] == year]
        df_year.to_excel(writer, sheet_name=str(year), index=False)

print(f"Planilha '{excel_file_name}' gerada com sucesso, com dados separados por ano em abas.")

Planilha 'EXP_Aeronautico_SP_com_Regiao_por_Ano.xlsx' gerada com sucesso, com dados separados por ano em abas.


## Summary:

### Data Analysis Key Findings

*   The initial merge of the export data with municipality data using original names failed for some entries due to inconsistencies in casing and accents.
*   Standardizing municipality names by removing accents and converting to uppercase before merging significantly improved the matching rate.
*   A small number of municipalities in the export data were not found in the provided municipality reference data and required manual mapping to their regions.
*   The final dataset includes a 'Região' column, successfully added by merging and handling missing values.
*   The export data spans multiple years, which were successfully used to create separate sheets in the output Excel file.

### Insights or Next Steps

*   Investigate the source of the municipality name discrepancies and missing entries to potentially improve data quality at the source.
*   Perform further analysis on the generated Excel file to identify trends in aeronautical exports by region and year.
